Add test to ensure existence of component docs

asuresh4 opened the pull request:
**Description:** Add a simple go test to ensure existence of documentation for enabled components.

**Link to tracking Issue:** #1025

asuresh4 commented:
FYI @tigrannajaryan @jrcamp

jrcamp commented:
I have a couple suggestions on possible simplifications, easier to show in a diff: https://github.com/jrcamp/opentelemetry-collector/commit/2863ce36946c8c7883707e3523de38f8593488a9

Ideally we can reuse this in contrib but I think some of the values will need to be parameterized. Are you planning to do those changes later or should we make it usable for both?

asuresh4 commented:
Thanks @jrcamp for the suggestions. Will incorporate those into this PR. As a first step I was planning on adding a similar test to the contrib repo. When you say reusable, where would the test live?

jrcamp commented:
> Thanks @jrcamp for the suggestions. Will incorporate those into this PR. As a first step I was planning on adding a similar test to the contrib repo. When you say reusable, where would the test live?

I think the test would live in core but it would have to be in a non-test package. Maybe see `componenttest` package as an example. There would have to be an actual test function in each but it can just call the shared package.

jrcamp commented on component/componenttest/component_docs.go:
```diff
@@ -0,0 +1,103 @@
+// getImportPrefixesToCheck returns a slice of strings that are relevant import
+// prefixes for components in the given module.
+func getImportPrefixesToCheck(module string) []string {
+	out := make([]string, len(componentTypes))
+	for i, typ := range componentTypes {
+		out[i] = fmt.Sprintf("%s/%s", strings.TrimRight(module, "/"), typ)
+	}
+	return out
```
Can you use `strings.Join`?

jrcamp commented:
lgtm, just one minor improvement suggested above. Looks like coverage isn't high enough yet though.

asuresh4 commented:
> lgtm, just one minor improvement suggested above. Looks like coverage isn't high enough yet though.

Addressed your comment. Still trying to get codecov/patch happy though.

tigrannajaryan commented on component/componentdocs/component_docs.go:
```diff
@@ -0,0 +1,103 @@
+	readMeFileName = "README.md"
+)
+
+// VerifyComponentDocumentation returns an error if README.md for at least one
+// enabled component is missing. "projectPath" is the absolute path to the root
+// of the project to which the components belong. "defaultComponentsFilePath" is
+// the path to the file that contains imports to all required components,
+// "goModule" is the Go module to which the imports belong. This method is intended
+// to be used only to verify documentation in Opentelemetry core and contrib
+// repositories. Examples,
+// 1) Usage in the core repo:
+// componenttest.VerifyComponentDocumentation(
+//		"path/to/project",
+//		"service/defaultcomponents/defaults.go",
+//      "go.opentelemetry.io/collector",
+//	)
+//
+// 2) Usage in the contrib repo:
+// componenttest.VerifyComponentDocumentation(
+//		"path/to/project",
+//		"cmd/otelcontrib/components.go",
+//      "github.com/open-telemetry/opentelemetry-collector-contrib",
+//	).
```
Nice comment!

tigrannajaryan commented on component/componentdocs/component_docs_test.go:
```diff
@@ -0,0 +1,177 @@
+		name    string
+		args    args
+		wantErr bool
+	}{
```
Good to see all these cases tested, thanks.
